### DISCUSSION TITLE
How to add a dotfile and make it machine specific?

### DISCUSSION THREAD

**Discussion by i2** [twpayne/chezmoi#2824] [General]
Hi everyone,

For example, say I already have a .wslconfig file as following:
```
[wsl2]
kernel = C:\\temp\\bzImage
```

Is there a more direct streamlined way to add this to my chezmoi repo?
Right now, from reading documents [here](https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/#use-completely-different-dotfiles-on-different-machines), it seems I have to do:
1. Manually create a file called "dot_wslconfig.tmpl" in source state with this content:
```
{{ if eq .chezmoi.os "windows" -}}
[wsl2]
kernel = C:\\temp\\bzImage
{{ end -}}
```
2. and then add and push this to my chezmoi repo on github.

This workflow seems a bit manual to me.

Intuitively what I would like to do to achieve this is:

`chezmoi add --autotemplate ~/.wslconfig`

and that would create the tmpl for me. Am I missing something?

Thanks!

**Comment by halostatue** [comment]
There’s a bit more that you would need to do so that you don’t generate such a file on non-Windows machines (with a templated `.chezmoiignore`).

There’s some thoughts for #2673 that include #2273 that will could help with this in the future.

**Comment by twpayne** [comment]
There is [section of the user guide dedicated to exactly your question](https://www.chezmoi.io/user-guide/manage-machine-to-machine-differences/).

In short, using `.chezmoiignore` to ignore `.wslconfg` on all non-Windows machines is probably the best fit for your use case.

**Comment by i2** [comment]
@twpayne Thanks for the response.

Two related question:

1. I would like to use exactly same config file for neovim on Windows and Linux, but it would complain about the line-endings. How do I solve this issue? I read about [line-endings](https://www.chezmoi.io/reference/templates/directives/#line-endings) but I don't know where to put this directive/template?

2. Up top in that page, it says **File-specific** `template options can be set using template directives ...` Is it possible to enforce `native` line-ending for all files (say it's not just neovim's init.vim, but a whole bunch of other configs that I share between Windows and Linux). How would I achieve this?

I'd appreciate your answer for both these situations. Thanks!

**Reply by twpayne** [reply]
> 1. I would like to use exactly same config file for neovim on Windows and Linux, but it would complain about the line-endings. How do I solve this issue? I read about [line-endings](https://www.chezmoi.io/reference/templates/directives/#line-endings) but I don't know where to put this directive/template?

You need to make your neovim config file a template and put

    chezmoi:template:line-endings=native

somewhere in the file.



> 2\. Up top in that page, it says **File-specific** `template options can be set using template directives ...` Is it possible to enforce `native` line-ending for all files (say it's not just neovim's init.vim, but a whole bunch of other configs that I share between Windows and Linux). How would I achieve this?

This is not possible, as enforcing native line endings for all files would corrupt binary files. You have to use the template/line-endings method for each file individually. Alternatively, there may be a way to use git's `core.autocrlf` setting to have git do the transformations for you, but I have not tried this. Note that most Windows programs are clever enough to be able to understand UNIX line endings, so it should only be necessary in a few cases.

**Reply by Lockszmith-GH** [reply]
How about a global setting for native line ending for all templates? Does that exist?

**Reply by i2** [reply]
> ```
> chezmoi:template:line-endings=native
> ```

Thanks, I feel stupid! Before asking here, I tried that but I missed a curly bracket! Solved!

> ```
> {{- chezmoi:template:line-endings=native -}}
> ```

**Comment by i2** [comment]
And final question:

Take `scoop` as an example. The way chezmoi manages its dotfile for me works fine across Windows and Linux, with a caveat. Let's take a look at my `scoop\config.json.tmpl` file in my Windows machine:

```
{{ if (and (eq .chezmoi.hostname "DESKTOP-GIQAJHB") (eq .chezmoi.os "windows")) -}}
{
    "last_update":  "2023-03-03T21:30:52.3632459-04:00",
    "scoop_branch":  "master",
    "scoop_repo":  "https://github.com/ScoopInstaller/Scoop"
}
{{ end -}}
```

When I do `chezmoi update` on my Linux, it will obviously not pull the config.json file, but it creates an empty scoop directory. How can I avoid this?

[ [this issue](https://github.com/twpayne/chezmoi/issues/754) might be related. ]

**Reply by halostatue** [reply]
```
# .chezmoiignore
{{ if eq .chezmoi.os "linux" -}}
scoop/
{{ end }}
```

I think. It’s been a long time since I’ve needed to worry about that, but it would be the same as ignoring the `.wslconfig`.

**Reply by bradenhilton** [reply]
Are you actually managing Scoop's config? Based on its contents in your example, you don't need to, as Scoop will create that minimum viable config automatically. You can verify this by deleting or renaming `config.json` and running `scoop update` a few times.

Regardless, I don't recommend managing Scoop's config with a regular template, as it updates the datetime in `last_update` frequently when it runs.

You should instead use a `modify_` script or template, like I have in my `modify_config.json`:

```text
// chezmoi:modify-template
// chezmoi:template:line-endings=crlf

{{- .chezmoi.stdin | fromJson |
    setValueAtPath "aria2-enabled" true |
    setValueAtPath "aria2-warning-enabled" false |
    setValueAtPath "use_external_7zip" true |
    toPrettyJson -}}
```

As @halostatue said, you should remove `if eq .chezmoi.os "windows"` and instead use `.chezmoiignore` to prevent the empty directory on Linux,

That said, it's slightly more correct to use

```text
{{ if ne .chezmoi.os "windows" }}
.config/scoop/
{{ end }}
```

as Scoop is only used on Windows.

**Reply by i2** [reply]
@bradenhilton I actually don't chezmoi scoop. It was just an example for a cross-platform program. I admit it was a poor choice :D

**Comment by i2** [comment]
I manage my dotfiles using a customized `git bare` method (I made quite a few changes in how the popular method available on the internet works).

I should mention that chezmoi does have quite a learning curve to get things right. It is a different mindset.

As counter-intuitive that it might sound, git bare method might be more automated, whereas chezmoi, **in exchange for some more manual prep-work**, gives amazing flexibility and programmability. There is quite a bit of work need to be done. You submit a version of config/dotfile you like and then modify them using templates etc... And what actually gets written to github VCS is not straightforward a one-on-one translation to the actual config/dotfiles (which makes sense).

I hope all this prepwork with chezmoi is worth it (and I believe it is).

**Reply by halostatue** [reply]
Chezmoi really shows its power in secrets management, IMO. I have *many* AWS keys that I import from 1Password, and there are other items that I store in 1Password that I would never want to commit to a repo even encrypted.

**Reply by i2** [reply]
@halostatue `halo stat-you`r dotfiles! Chezmoi comes back after a few weeks and it still looks good and idempotent :)
